### PR TITLE
Widget class

### PR DESCRIPTION
**Pull request by vcoindet**
Adding Widget class in example

Widget class with constructor and method (Widget.js)

In Widget.js
- getName()
- getElement()
- getOptions()
- setOptions(pOptions)
- listenToMap(pEventName, pCallback)
- getGlobe()
- setGlobe(map)

Method in widgetAPI.js
- addWidget(widget)
- getWidgets()
- removeWidget(widget)

To make the examples work : put the following lines in index.html :
- < link rel="stylesheet" href="examples/widgets/widget.css" > in the 'head' tag
- <script src="examples/widgets/pickPositionWidget.js"> </script>
  <script src="examples/widgets/rangeWidget.js"> </script> 
  Right before the closure of the 'body' tag
- <script src="examples/widgets/Widget.js"> </script>
  <script src="examples/widgets/widgetAPI.js"> </script>
  Just after the line <script src="dist/itowns.js"></script>
- const widgetTab = new widgetAPI();

### CODE REVIEW

**Comment by vcoindet** [comment]
rebased and ready

File: examples/controls/control.css [hunk 1]
```diff
@@ -0,0 +1,12 @@
+#controlDiv {
+      position:absolute;
+      top:10px;
+      right:10px;
+      width:150px;
+      height:150px;
```
Comment (autra) [inline examples/controls/control.css]: nits: space before `:`

File: src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js [hunk 2]
```diff
@@ -852,6 +857,45 @@ ApiGlobe.prototype.removeEventListenerLayerChanged = function removeEventListene
     this.viewerDiv.removeEventListener('layerchanged:index', this.callbackLayerChanged, false);
 };
 
+/**
+ * Add a control.
+ * @constructor
```
Comment (autra) [inline src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js]: Not a constructor! (and below too)

File: src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js [hunk 3]
```diff
@@ -852,6 +857,45 @@ ApiGlobe.prototype.removeEventListenerLayerChanged = function removeEventListene
...
+ * @param {object} Control - The Control object.
+ */
+
+ApiGlobe.prototype.addControl = function addControl(control) {
+    var index = this.controls.indexOf(control);
```
Comment (autra) [inline src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js]: Maybe a Set is a more natural data structure here.

File: src/Core/Commander/Interfaces/Control/Control.js [hunk 4]
```diff
@@ -0,0 +1,88 @@
+/**
+ * Generated On: 2016-11-22
+ * Class: Controls
+ * Description: Cette Classe construit un control.
```
Comment (autra) [inline src/Core/Commander/Interfaces/Control/Control.js]: Should be in English

**Comment by peppsac** [comment]
@vcoindet can you please squash or reword your commits? (5 commits use the same text)

File: src/Core/Commander/Interfaces/Control/Control.js [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+
+function Control(pName, pElement, options) {
+    this.pName = pName;
+    this.pElement = pElement;
+    this.options = options;
+    this.map_ = null;
+    this.api = ApiGlobe;
+}
+
+Control.prototype.constructor = Control;
```
Comment (autra) [inline src/Core/Commander/Interfaces/Control/Control.js]: you don't need this if you don't overwrite the `prototype`.

File: src/Core/Commander/Interfaces/Control/Control.js [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+ * Description: Cette Classe construit un control.
+ */
+
+import ApiGlobe from 'Core/Commander/Interfaces/ApiInterface/ApiGlobe';
+
+function Control(pName, pElement, options) {
+    this.pName = pName;
+    this.pElement = pElement;
+    this.options = options;
+    this.map_ = null;
```
Comment (autra) [inline src/Core/Commander/Interfaces/Control/Control.js]: Is it supposed to be a private field? In this case, the convention is to name it `this._map`.

File: src/Core/Commander/Interfaces/Control/Control.js [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+ * @constructor
+ * @param {string} Eventname - The name of the event.
+ * @param {callback} Callback - The callback that is called when the event is heard.
+ */
+
+Control.prototype.listenToMap = function listenToMap(pEventName, pCallback) {
+    document.getElementById('viewerDiv').addEventListener(pEventName, pCallback, false);
+};
+
+Control.prototype.getGlobe = function getGlobe() {
```
Comment (autra) [inline src/Core/Commander/Interfaces/Control/Control.js]: Might not be a globe in all cases (we want to make the geometry configurable in the future right?)

File: src/Core/Commander/Interfaces/Control.js [hunk 8]
```diff
@@ -0,0 +1,67 @@
+/**
```
Comment (autra) [inline src/Core/Commander/Interfaces/Control.js]: One of `src/Core/Commander/Interfaces/Control.js` and `src/Core/Commander/Interfaces/Control/Control.js` should be removed I guess.

File: index.html [hunk 9]
```diff
@@ -42,6 +53,7 @@
             // iTowns namespace defined here
             var viewerDiv = document.getElementById('viewerDiv');
             var menuGlobe = new GuiTools(itowns.viewer,'menuDiv');
+
```
Comment (autra) [inline index.html]: Not very critical, but in general it's best to remove unrelated changes from a PR. (And submit them elsewhere if needed).

File: index.html [hunk 10]
```diff
@@ -68,7 +80,7 @@
 
             Promise.all([imageryPromise, elevationPromise]).then(() => itowns.viewer.init());
 
-            menuGlobe.addGUI('RealisticLighting',false,function(newValue) {itowns.viewer.setRealisticLightingOn(newValue);});
+            menuGlobe.addGUI('RealisticLighting',false,(newValue) => {itowns.viewer.setRealisticLightingOn(newValue);});
```
Comment (autra) [inline index.html]: Same remark, unrelated change.

File: src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js [hunk 11]
```diff
@@ -660,6 +663,7 @@ ApiGlobe.prototype.setRange = function setRange(pRange /* , bool anim*/) {
     eventRange.oldRange = this.getRange();
     loaded = false;
     this.scene.currentControls().setRange(pRange);
+    this.viewerDiv.dispatchEvent(eventRange);
```
Comment (autra) [inline src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js]: Be careful, events should not be used as static variable. THe browser sets some fields (`targets`, `innerTargets`). By using the same ref for every events, you risk some collision. It's best to create a new instance at each `dispatch`.

File: src/Core/Commander/Interfaces/Control/Control.js [hunk 12]
```diff
@@ -0,0 +1,88 @@
...
+    return this.map_;
+};
+
+Control.prototype.setGlobe = function setGlobe(map) {
+    if (this.map_) {
+        document.getElementById(this.options.div.id).removeChild(this.pElement);
```
Comment (autra) [inline src/Core/Commander/Interfaces/Control/Control.js]: If I'm not mistaken `this.options.div` is already the same dom element returned by `document.getElementById(this.options.div.id)`.

File: src/Core/Commander/Interfaces/Control/Control.js [hunk 13]
```diff
@@ -0,0 +1,88 @@
...
+    }
+    this.map_ = map;
+    if (this.map_) {
+        var element = this.pElement;
+        document.getElementById(this.options.div.id).appendChild(element);
```
Comment (autra) [inline src/Core/Commander/Interfaces/Control/Control.js]: same here.

File: src/Core/Commander/Interfaces/Control/Control.js (hunk 7 above)
Comment (autra) [reply src/Core/Commander/Interfaces/Control/Control.js]: `getMap` and `setMap`?

File: src/Core/Commander/Interfaces/Control/Control.js [hunk 14]
```diff
@@ -0,0 +1,88 @@
...
+};
+
+/**
+ * Return the options of the control.
+ * @constructor
+ * @return     {object}  Object.
+ */
+
+Control.prototype.getOptions = function getOptions() {
+    this.options = this.options || '';
```
Comment (autra) [inline src/Core/Commander/Interfaces/Control/Control.js]: you meant `this.options = this.options || {};` right?

File: src/Core/Commander/Interfaces/Control/Control.js [hunk 15]
```diff
@@ -0,0 +1,88 @@
...
+
+/**
+ * Listen to an event linked to the globe.
+ * @constructor
+ * @param {string} Eventname - The name of the event.
+ * @param {callback} Callback - The callback that is called when the event is heard.
+ */
+
+Control.prototype.listenToMap = function listenToMap(pEventName, pCallback) {
+    document.getElementById('viewerDiv').addEventListener(pEventName, pCallback, false);
```
Comment (autra) [inline src/Core/Commander/Interfaces/Control/Control.js]: there should be a `removeEventListener` somewhere, as apparently you can remove controls. Failure to do so would prevent the browser to dispose of your object.

File: examples/controls/control.css (hunk 1 above)
Comment (autra) [inline examples/controls/control.css]: Consider using `rem` units instead of px to accomodate with different screen size and density.

File: examples/controls/control.css [hunk 16]
```diff
@@ -0,0 +1,12 @@
...
+      background-color: white;
+      border-style: solid;
+      border-color: black;
+      border-width: 1px;
+      box-shadow: 1px 1px 10px #555;
```
Comment (autra) [inline examples/controls/control.css]: (for border and boxes, some people prefer staying with px, they think it's nicer)

File: examples/controls/control.css [hunk 17]
```diff
@@ -0,0 +1,12 @@
+#controlDiv {
```
Comment (autra) [inline examples/controls/control.css]: This file should be called `controls.css`, and this div `controlsDiv`. Or even better `#controls`. The fact it is a div is an implementation detail.

File: examples/controls/control.css [hunk 18]
```diff
@@ -0,0 +1,12 @@
...
+      top:10px;
+      right:10px;
+      width:150px;
+      height:150px;
+      background-color: white;
+      border-style: solid;
+      border-color: black;
+      border-width: 1px;
+      box-shadow: 1px 1px 10px #555;
+}
```
Comment (autra) [inline examples/controls/control.css]: I actually suggest:

```css
position: absolute;
top: 1rem;
right: 1rem;
width: 20rem;
box-shadow: 1px 1px 10px #555;
font: 11px Lucida Grande,sans-serif;
color: #eee;
padding: 0.5rem;
```

And separating the fields with `<br>` instead of `;` in your callbacks (in `pickposition_controls`).

Result:

screen shot 2017-01-16 at 15 22 27

File: examples/controls/pickPosition_control.js [hunk 19]
```diff
@@ -0,0 +1,21 @@
+var pickControl = document.createElement('div');
```
Comment (autra) [inline examples/controls/pickPosition_control.js]: For the filename: either choose snake_case or CamelCase. (the rest of the project use CamelCase if I'm not mistaken).

File: examples/controls/pickPosition_control.js [hunk 20]
```diff
@@ -0,0 +1,21 @@
...
+
+var controlDiv = document.getElementById('controlDiv');
+
+controlDiv.appendChild(pickControl);
+
+var control = new itowns.viewer.Control(
+    'Pick',
+    document.getElementById('pickControl'),
+    {
+        div: document.getElementById('controlDiv')
```
Comment (autra) [inline examples/controls/pickPosition_control.js]: `div: controlDiv`. You have it just above.

File: examples/controls/pickPosition_control.js [hunk 21]
```diff
@@ -0,0 +1,21 @@
...
+    }
+);
+
+var callback = function () {
+    itowns.viewer.addControl(control);
+    document.getElementById(control.pElement.id).innerHTML = "longitude : " + itowns.viewer.pickPosition().coordinate[0] * 180 / 3.14 + " ; latitude : " + itowns.viewer.pickPosition().coordinate[1] * 180 / 3.14 + " ; altitude : " +  itowns.viewer.pickPosition().coordinate[2];
+};
+
+itowns.viewer.addEventListener('mousedown', callback);
```
Comment (autra) [inline examples/controls/pickPosition_control.js]: So I don't understand: how is `Control` useful if you do all the logic outside of it? You're not even using `listenToMap`... (though I think a callback in the `options` constructor parameter would be a better idea than listenToMap).

File: src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js [hunk 22]
```diff
@@ -852,6 +857,45 @@ ApiGlobe.prototype.removeEventListenerLayerChanged = function removeEventListene
...
+ * @constructor
+ * @param {object} Control - The Control object.
+ */
+
+ApiGlobe.prototype.removeControl = function removeControl(control) {
+    var index = this.controls.indexOf(control);
+    if (index > -1) {
+        this.controls.splice(index, 1);
+    }
+    control.setGlobe();
```
Comment (autra) [inline src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js]: Let's be explicit `control.setGlobe(null);`

But it sounds like an abuse to me. I would prefer having a `control.dispose` method that would remove listeners and stuff.

File: index.html [hunk 23]
```diff
@@ -68,7 +80,7 @@
...
 
            	itowns.viewer.addEventListener('globe-loaded', function() {
             	console.log("Globe Loaded");
```
Comment (autra) [inline index.html]: Actually, we could have several example html, one basic like it is now, another with all the controls etc... WDYT?

File: src/Core/Commander/Interfaces/Control/Control.js [hunk 24]
```diff
@@ -0,0 +1,88 @@
+/**
+ * Generated On: 2016-11-22
+ * Class: Controls
+ * Description: Cette Classe construit un control.
+ */
+
+import ApiGlobe from 'Core/Commander/Interfaces/ApiInterface/ApiGlobe';
```
Comment (autra) [inline src/Core/Commander/Interfaces/Control/Control.js]: Not used anywhere. Also, cyclic dependencies should be avoided as much as possible (it creates difficult-to-maintain code).

**Review by autra — COMMENTED** [review]
As nobody was looking at your PR, I took upon myself to do it :-) 

Can you explain to me what is the exact role of `Controls`? Is it to display information from some events? if so, maybe we can find another name (controls is for controlling stuff. We're not really controlling them here). `indicators`? `sensors`?

Also, my general feeling about this is that it is quite a rigid way to give information to client code. Don't we just want some events (optionally embedding useful informations, like new values) and let client code (geoportal, anything) deal with that how they see fit? 

As it is now, if I just want to store range info somewhere (for history or telemetry purpose for example), I cannot do that without displaying them. But maybe I didn't get the rationale here.

File: examples/controls/control.css (hunk 17 above)
Comment (vcoindet) [reply examples/controls/control.css]: OK I agree

File: examples/controls/control.css (hunk 1 above)
Comment (vcoindet) [reply examples/controls/control.css]: Will work on that

File: examples/controls/control.css (hunk 18 above)
Comment (vcoindet) [reply examples/controls/control.css]: I don't usually use `<br>` for semantic reasons. I'd rather use `<li>`. I agree that `;` wasn't a good solution, but it was just an example to test the control class

File: src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js (hunk 3 above)
Comment (vcoindet) [reply src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js]: You're probably right, didn't know this method at the time. Will change that. Or Maybe a `new Map()` can do the job too.

File: src/Core/Commander/Interfaces/Control/Control.js (hunk 24 above)
Comment (vcoindet) [reply src/Core/Commander/Interfaces/Control/Control.js]: You're absolutely right, I forget to remove it

File: src/Core/Commander/Interfaces/Control/Control.js (hunk 5 above)
Comment (vcoindet) [reply src/Core/Commander/Interfaces/Control/Control.js]: OK

File: src/Core/Commander/Interfaces/Control/Control.js (hunk 12 above)
Comment (vcoindet) [reply src/Core/Commander/Interfaces/Control/Control.js]: Well, I'm not sure of that. I have to test it, I will let you know

File: src/Core/Commander/Interfaces/Control.js (hunk 8 above)
Comment (vcoindet) [reply src/Core/Commander/Interfaces/Control.js]: Oups, my mistake, the first one has to be removed...

File: examples/controls/pickPosition_control.js (hunk 19 above)
Comment (vcoindet) [reply examples/controls/pickPosition_control.js]: lowerCamelCase if I'm not mistaken too. The best is that i choose lowerCamelCase.

File: examples/controls/pickPosition_control.js (hunk 20 above)
Comment (vcoindet) [reply examples/controls/pickPosition_control.js]: absolutely !

File: examples/controls/control.css (hunk 18 above)
Comment (autra) [reply examples/controls/control.css]: yep :-) Indeed, `li` are even better :-)

File: src/Core/Commander/Interfaces/Control/Control.js (hunk 7 above)
Comment (vcoindet) [reply src/Core/Commander/Interfaces/Control/Control.js]: Yep, I guess you're right, I hesitated between those two choices, but I only thought about 3D and not 2D

File: examples/controls/pickPosition_control.js (hunk 19 above)
Comment (autra) [reply examples/controls/pickPosition_control.js]: Actually yes, lowerCamelCase, thanks for translating for me ;-)

**Comment by vcoindet** [comment]
To be clear on the purpose of the Controls class : 
A control is a widget with a DOM element which position should be fixed on the screen (determined by using CSS) . It can involve inputs or be informational only.
You can use it for simple custom controls by creating the element with listeners, for example displaying the IPRs on the screen : IPRs can change when you remove a layer (event `layerchanged:visible`) or when you change the center of the map.

Here the control class is inspired on the OpenLayers 3's control class.

All the methods implemented in the Controls class can be used to help you display your control on the screen. So basically, it is used to create DOM elements and all controls inherits of this class.
Some examples of controls : menu, IPR, scale bar, mini Globe, zoom and dezoom button, etc... (see the screenshot for IPRs below)
ipr

File: index.html (hunk 23 above)
Comment (vcoindet) [reply index.html]: I totally agree, as controls like the menu, IPR, range, picking, etc... should be examples.

File: src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js (hunk 11 above)
Comment (vcoindet) [reply src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js]: So I guess we should change that for all dispatched events. Maybe in an another PR ? WDYT ?

**Comment by autra** [comment]
Thanks for the context !

So actually, those Controls would exactly do dat.gui's job + displaying read-only informations, right? If I understand it correctly, why don't we use dat.js for this? It's already a very good state-of-the-art lib, with more than 1000 stars on github... and we are already using it! So it would mean less maintenance for us, and a better integration with the currently enabled controls. 

For example, with the current master, with a small modification to GuiTools.js (return the result of `this.gui..add` on line 74), I can achieve a range spy just with these following lines in index.html:
```js
let myUi = menuGlobe.addGUI('range', {itowns.viewer.getRange()}, (newValue) => { /* do nothing */ });
myUi.domElement.children[0].disabled = true; // make this read-only
itowns.viewer.addEventListener('rangeChanged', () => {
    console.log('new range', itowns.viewer.getRange());
    menuGlobe.range = itowns.viewer.getRange();
    myUi.updateDisplay();
});
```
So with some polishing (creating a new method in GuiTools or modifiying `addGUI` by adding a `readonly` boolean parameter and an event name to listen to), we already have everything we need, no need to create another module for that (+ some css etc).

Note that the example above only works if you manually execute `itowns.viewer.setRange(...)` in the console, but that's because we don't correctly fire `rangechanged` events when we do it from GlobeControls.js. So this needs to get fixed anyway (rangechanged events should be fired from GlobeControls I guess).

So to summarize, I thing that:
- we should fix our events in ApiGlobe and ensure they are fired when they should.
- we should use dat.gui for these controls.

One last thing:

> A control is a widget

So let's call them Widgets? Because for us, controls are a different thing (mouse and keyboard actions), ex GlobeControls.js (and I hope to make them configurable and to have FPSControls, PanoramicControls etc...).

File: src/Core/Commander/Interfaces/Control/Control.js (hunk 12 above)
Comment (vcoindet) [reply src/Core/Commander/Interfaces/Control/Control.js]: Yep, you're right, so I changed it

File: src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js (hunk 11 above)
Comment (autra) [reply src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js]: yep, you're right :-)

**Comment by vcoindet** [comment]
Commit squashed, Rebased and Ready
Changed the words 'controls' by 'widgets' as suggested by @autra

**Comment by vcoindet** [comment]
Rebased and Ready

Put everything in examples, please read the description of this PR to make it work correctly.

File: .gitattributes [hunk 25]
```diff
@@ -1,4 +1,4 @@
-*	text=auto
+*   text=auto
```
Comment (gchoqueux) [inline .gitattributes:1]: reason?

File: examples/widgets/Widget.js [hunk 26]
```diff
@@ -0,0 +1,104 @@
...
+};
+
+/**
+ * Associate a map to a widget.
+ * @constructor
+ */
+
+Widget.prototype.setMap = function setMap(map) {
+    if (this._map) {
+        this.options.div.removeChild(this.pElement);
```
Comment (gchoqueux) [inline examples/widgets/Widget.js:95]: why remove and re-add this.pElement?

File: examples/widgets/Widget.js [hunk 27]
```diff
@@ -0,0 +1,104 @@
...
+    }
+    this._map = map;
+    if (this._map) {
+        var element = this.pElement;
```
Comment (gchoqueux) [inline examples/widgets/Widget.js:99]: remove `var element` and replace `this.options.div.appendChild(element);` by `this.options.div.appendChild(this.pElement);`

File: examples/widgets/pickPositionWidget.js [hunk 28]
```diff
@@ -0,0 +1,37 @@
+/* global itowns */
```
Comment (gchoqueux) [inline examples/widgets/pickPositionWidget.js:1]: rename by PickPositionWidget.js

File: examples/widgets/Widget.js [hunk 29]
```diff
@@ -0,0 +1,104 @@
+/**
+ * Generated On: 2016-11-22
+ * Class: Controls
+ * Description: Create a widget.
+ */
+
+function Widget(pName, pElement, options) {
```
Comment (gchoqueux) [inline examples/widgets/Widget.js:7]: export Widget

File: examples/widgets/pickPositionWidget.js [hunk 30]
```diff
@@ -0,0 +1,37 @@
...
+/* global Widget */
```
Comment (gchoqueux) [inline examples/widgets/pickPositionWidget.js:2]: import Widget

File: examples/widgets/pickPositionWidget.js [hunk 31]
```diff
@@ -0,0 +1,37 @@
...
+/* global widgetTab */
```
Comment (gchoqueux) [inline examples/widgets/pickPositionWidget.js:3]: remove global  widgetTab 

User must do widgetTab.add(pickPositionWidget);

Do example to test pickPositionWidget

File: examples/widgets/pickPositionWidget.js [hunk 32]
```diff
@@ -0,0 +1,37 @@
...
+        div: widgetsP,
+    });
+
+const callback = function () {
+    widgetTab.addWidget(widget);
+    if (document.getElementById('pickList')) {
+        document.getElementById(widget.pElement.id).removeChild(document.getElementById('pickList'));
+    }
+    const tab = ['longitude', 'latitude', 'altitude'];
+    const coords = itowns.viewer.pickPosition().coordinate;
```
Comment (gchoqueux) [inline examples/widgets/pickPositionWidget.js:25]: remove global api call, pass in api in constructor

File: examples/widgets/rangeWidget.js [hunk 33]
```diff
@@ -0,0 +1,31 @@
+/* global itowns */
```
Comment (gchoqueux) [inline examples/widgets/rangeWidget.js:1]: no global variable and use import export, see comment for pick

rename file by RangeWidget.js

Comment (gchoqueux) [reply examples/widgets/rangeWidget.js:1]: add example

File: examples/widgets/widgetAPI.js [hunk 34]
```diff
@@ -0,0 +1,34 @@
+/* global Widget */
+/* global itowns */
+
+function widgetAPI() {
```
Comment (gchoqueux) [inline examples/widgets/widgetAPI.js:4]: widgetApi is not really necessary

**Review by gchoqueux — CHANGES_REQUESTED** [review]

File: examples/widgets/pickPositionWidget.js [hunk 35]
```diff
@@ -0,0 +1,37 @@
...
+    const list = document.createElement('ul');
+    list.id = 'pickList';
+    for (var j = 0; j < coords.length; j++) {
```
Comment (gchoqueux) [inline examples/widgets/pickPositionWidget.js:28]: replace by for ... in

Comment (gchoqueux) [reply examples/widgets/pickPositionWidget.js:28]: sorry replace by for ... of

File: examples/widgets/Widget.js (hunk 29 above)
Comment (autra) [reply examples/widgets/Widget.js:7]: I might be late to the party, but you cannot really do that: the examples are neither handled by webpack nor by babel. Imho they should be seen as ready-to-use snippets of code for client apps. Therefore, they shouldn't rely on webpack, babel or anything linked to our build system. They should be valid, browser-ready javascript for our target browsers.

This actually means that:
- we need to formally decide which our target browsers (and versions) are (because it defines which js api/functions/syntax we can safely use)
- we need a special `.eslintrc` for the `examples` folder

File: examples/widgets/pickPositionWidget.js (hunk 35 above)
Comment (autra) [reply examples/widgets/pickPositionWidget.js:28]: (if for..of is supported by our target browsers, we won't use babel here I guess)

File: examples/widgets/Widget.js (hunk 29 above)
Comment (vcoindet) [reply examples/widgets/Widget.js:7]: Sorry for the late answer... You're not too late ! As you mentionned it, it's not possible to do that in `examples`.
In fact, I thought maybe we could do a new `index.html` in `/examples/` for each widget's examples, but it won't solve the problem.
I agree with your solution, I think it's the best we can think about right now. But I don't have any knowledge on `.eslintrc`, so do you agree to help me to do a special `.eslintrc` for the `examples` folder ?
